core/shared/src/main/scala/zio/InMemoryCache.scala

```diff
@@ -0,0 +1,49 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.annotation.tailrec
+
+final class InMemoryCache[E, K, V] private (private val ref: AtomicReference[Map[K, Promise[E, V]]])
```

Comment:
I see this implementation is very close to the `Promise`, or `Ref` implementations. I wonder if this is a higher level concept and could be represented in terms of some other, lower level primitive.